frontend/model/state.js

```diff
@@ -92,26 +92,65 @@ sbp('sbp/selectors/register', {
-          return upgradeRequired || (member.status === PROFILE_STATUS.ACTIVE && member.joinedHeight == null)
+        }).reduce((contractID: string | boolean, member: Object) => {
+          if (contractID) return contractID
+          if (member.status === PROFILE_STATUS.ACTIVE && member.joinedHeight == null) {
+            return groupID
+          }
+          return false
         }, false)
+      }).forEach((contractID) => {
+        if (!contractID) return
+        sbp('gi.actions/group/upgradeFrom1.0.7', { contractID }).catch(e => {
+          console.error('[state/vuex/postUpgradeVerification] Error during gi.actions/group/upgradeFrom1.0.7', contractID, e)
+        })
       })
-      if (!upgradeRequired) return
-      sbp('gi.actions/group/upgradeFrom1.0.7').catch(e => {
-        console.error('[state/vuex/postUpgradeVerification] Error during gi.actions/group/upgradeFrom1.0.7', e)
-      })
+    })();
+    (() => {
+      // Upgrade from version 1.0.8 to a newer version
+      // The new chatroom contracts have an admin IDs list
+      // This code checks if the attribute is missing, and if so, issues the
+      // corresponing upgrade action.
+      const ourIdentityContractId = state.loggedIn?.identityContractID
+      if (!ourIdentityContractId || !state[ourIdentityContractId]) return
+      if (state[ourIdentityContractId].groups) {
+        // Group chatrooms
+        Object.entries(state[ourIdentityContractId].groups).map(([groupID, { hasLeft }]: [string, Object]) => {
+          if (hasLeft || !state[groupID]?.chatRooms || !state[groupID].groupOwnerID) return []
+          // $FlowFixMe[incompatible-use]
+          return Object.entries((state[groupID].chatRooms: { [string]: Object })).flatMap(([chatroomID, { members }]) => {
+            return members[ourIdentityContractId]?.status === PROFILE_STATUS.ACTIVE && !Array.isArray(state[chatroomID]?.attributes?.adminIDs) && [chatroomID, state[groupID].groupOwnerID]
+          })
+        }).forEach(([contractID, groupOwnerID]) => {
+          if (!contractID) return
+          sbp('gi.actions/chatroom/upgradeFrom1.0.8', { contractID, data: groupOwnerID }).catch(e => {
+            console.error('[state/vuex/postUpgradeVerification] Error during gi.actions/chatroom/upgradeFrom1.0.8', contractID, e)
+          })
+        })
+      }
+      if (state[ourIdentityContractId].chatRooms) {
+        // DM chatrooms
+        return Object.keys((state[ourIdentityContractId].chatRooms: { [string]: Object })).map((chatroomID) => {
+          return state[chatroomID]?.members[ourIdentityContractId] && !Array.isArray(state[chatroomID].attributes?.adminIDs) && chatroomID
+        }).forEach((contractID) => {
+          if (!contractID) return
+          sbp('gi.actions/chatroom/upgradeFrom1.0.8', { contractID }).catch(e => {
+            console.error('[state/vuex/postUpgradeVerification] Error during gi.actions/chatroom/upgradeFrom1.0.8', contractID, e)
+          })
+        })
+      }
```

Comment:
I ran this logic through Claude Sonnet 3.5 and it suggested this:

```js
if (state[ourIdentityContractId]?.groups) {
  const upgradeTasks = Object.entries(state[ourIdentityContractId].groups)
    .flatMap(([groupID, { hasLeft }]) => {
      if (hasLeft || !state[groupID]?.chatRooms || !state[groupID]?.groupOwnerID) return []
      
      return Object.entries(state[groupID].chatRooms)
        .filter(([chatroomID, { members }]) => 
          members[ourIdentityContractId]?.status === PROFILE_STATUS.ACTIVE && 
          !Array.isArray(state[chatroomID]?.attributes?.adminIDs)
        )
        .map(([chatroomID]) => ({ contractID: chatroomID, groupOwnerID: state[groupID].groupOwnerID }))
    })

  for (const { contractID, groupOwnerID } of upgradeTasks) {
    try {
      await sbp('gi.actions/chatroom/upgradeFrom1.0.8', { contractID, data: groupOwnerID })
    } catch (e) {
      console.error(`[state/vuex/postUpgradeVerification] Error during gi.actions/chatroom/upgradeFrom1.0.8 for ${contractID}:`, e)
      // Consider adding more robust error handling here
    }
  }
}
```

Is there anything to that or is it nonsense?

Reply:
I don't see anything obviously wrong with it at a glance, but I'm not sure what the point of this is. Was it for it to be clearer / better organised? If so, I think the `upgradeAction` and `needsUpgrade` functions make the intent clearer and is slightly less repetitive.

Reply:
> Was it for it to be clearer / better organised?

Yes, and DRYs it a bit.

Reply:
I've taken some of the suggestions.